File: masker/voice_loop.py
"""End-to-end voice / text loop. CURSOR OWNS THIS FILE.

Orchestrates: input → STT → detection → policy → masking → routing → output filter.
Returns a TurnResult that Ona's UI consumes for the trace view.

Two entry surfaces:

  - `run_text_turn(text)`   — headless, used by tests and the demo CLI
  - `run_voice_turn()`      — mic in / TTS out, calls into `cactus transcribe`
                              under the hood (Phase-1 uses scripted text fallback)
"""

from __future__ import annotations

import shutil
import subprocess
import time
from dataclasses import dataclass

from .contracts import PolicyName, TurnResult
from . import detection as _detection
from . import masking as _masking
from . import policy as _policy
from .router import Router, default_router
from .trace import Tracer


@dataclass
class VoiceLoop:
    router: Router
    policy_name: PolicyName = "hipaa_base"
    mask_mode: str = "placeholder"

    def run_text_turn(self, text: str, tracer: Tracer | None = None) -> TurnResult:
        tracer = tracer or Tracer()
        t0 = time.perf_counter()

        with tracer.span("detection", "Scanning input for PII/PHI"):
            det = _detection.detect(text)
        tracer.event(
            "detection",
            f"risk={det.risk_level}, entities={len(det.entities)}",
            risk=det.risk_level,
            entity_types=[e.type.value for e in det.entities],
        )

        with tracer.span("policy", f"Applying {self.policy_name}"):
            decision = _policy.decide(det, policy_name=self.policy_name)
        tracer.event("policy", f"route={decision.route}", **decision.to_dict())

        with tracer.span("masking", "Masking sensitive spans"):
            masked = _masking.mask(text, det, mode=self.mask_mode)  # type: ignore[arg-type]
        if masked.token_map:
            tracer.event(
                "masking",
                f"masked {len(masked.token_map)} span(s)",
                masked_count=len(masked.token_map),
            )

        model_text = self.router.execute(
            original_text=text,
            masked=masked,
            decision=decision,
            tracer=tracer,
        )

        with tracer.span("output_filter", "Re-scanning model output for leakage"):
            safe_out = _masking.scrub_output(model_text, det)
        if safe_out != model_text:
            tracer.event("output_filter", "scrubbed leaked entity from output")

        total_ms = (time.perf_counter() - t0) * 1000.0
        return TurnResult(
            user_text=text,
            detection=det,
            policy=decision,
            masked_input=masked,
            model_output=model_text,
            safe_output=safe_out,
            trace=list(tracer.events),
            total_ms=total_ms,
        )

    def run_voice_turn(self, *, audio_file: str | None = None, tracer: Tracer | None = None) -> TurnResult:
        """Voice-in variant. Uses `cactus transcribe` to convert audio → text.
        Falls back to `input()` prompt if no `cactus` CLI is available so dev
        loops still work without a mic / model.
        """
        tracer = tracer or Tracer()
        with tracer.span("stt", "Transcribing audio"):
            text = _transcribe(audio_file)
        if not text:
            text = input("masker> ").strip()
        return self.run_text_turn(text, tracer=tracer)


def _transcribe(audio_file: str | None) -> str:
    if not shutil.which("cactus"):
        return ""
    cmd = ["cactus", "transcribe"]
    if audio_file:
        cmd += ["--file", audio_file]
    try:
        proc = subprocess.run(cmd, capture_output=True, text=True, timeout=30)
    except subprocess.TimeoutExpired:
        return ""
    if proc.returncode != 0:
        return ""
    return proc.stdout.strip()


def default_loop(*, policy_name: PolicyName = "hipaa_base") -> VoiceLoop:
    return VoiceLoop(router=default_router(), policy_name=policy_name)

File: masker-rs/crates/masker/src/voice_loop.rs
//! End-to-end voice / text loop. Mirrors `masker/voice_loop.py`.

use std::time::Instant;

use crate::contracts::{PolicyName, TraceStage, TurnResult};
use crate::masking::MaskMode;
use crate::payload;
use crate::router::Router;
use crate::trace::Tracer;
use crate::{detection, masking, policy};

pub struct VoiceLoop {
    pub router: Router,
    pub policy_name: PolicyName,
    pub mask_mode: MaskMode,
}

impl VoiceLoop {
    pub fn new(router: Router) -> Self {
        Self {
            router,
            policy_name: PolicyName::HipaaBase,
            mask_mode: MaskMode::Placeholder,
        }
    }

    pub fn with_policy(mut self, p: PolicyName) -> Self {
        self.policy_name = p;
        self
    }

    pub fn with_mask_mode(mut self, m: MaskMode) -> Self {
        self.mask_mode = m;
        self
    }

    /// Run a full text turn — used by tests, CLI demo, and any integration
    /// that already has transcribed text.
    pub fn run_text_turn(&self, text: &str, tracer: &Tracer) -> TurnResult {
        let t0 = Instant::now();

        let detection = {
            let _s = tracer.span(
                TraceStage::Detection,
                "Scanning input for PII/PHI",
                payload! {},
            );
            detection::detect(text)
        };
        tracer.event(
            TraceStage::Detection,
            format!(
                "risk={}, entities={}",
                detection.risk_level.as_str(),
                detection.entities.len()
            ),
            payload! {
                "risk" => detection.risk_level.as_str(),
                "entity_types" => detection.entities.iter().map(|e| e.kind.as_str()).collect::<Vec<_>>(),
            },
        );

        let decision = {
            let _s = tracer.span(
                TraceStage::Policy,
                format!("Applying {}", self.policy_name.as_str()),
                payload! {},
            );
            policy::decide(&detection, self.policy_name)
        };
        tracer.event(
            TraceStage::Policy,
            format!("route={}", decision.route.as_str()),
            payload! {
                "route"     => decision.route.as_str(),
                "policy"    => decision.policy.as_str(),
                "rationale" => decision.rationale.clone(),
            },
        );

        let masked = {
            let _s = tracer.span(
                TraceStage::Masking,
                "Masking sensitive spans",
                payload! {},
            );
            masking::mask(text, &detection, self.mask_mode)
        };
        if !masked.token_map.is_empty() {
            tracer.event(
                TraceStage::Masking,
                format!("masked {} span(s)", masked.token_map.len()),
                payload! { "masked_count" => masked.token_map.len() },
            );
        }

        let model_text = self.router.execute(text, &masked, &decision, tracer);

        let safe_out = {
            let _s = tracer.span(
                TraceStage::OutputFilter,
                "Re-scanning model output for leakage",
                payload! {},
            );
            masking::scrub_output(&model_text, &detection)
        };
        if safe_out != model_text {
            tracer.event(
                TraceStage::OutputFilter,
                "scrubbed leaked entity from output",
                payload! {},
            );
        }

        let total_ms = t0.elapsed().as_secs_f64() * 1000.0;

        TurnResult {
            user_text: text.to_string(),
            detection,
            policy: decision,
            masked_input: masked,
            model_output: model_text,
            safe_output: safe_out,
            trace: tracer.events(),
            total_ms,
        }
    }
}

pub fn default_loop() -> VoiceLoop {
    VoiceLoop::new(crate::router::default_router())
}
